src/content/post/2024-04-05-rdf-shacl-and-kuzu.md

```diff
@@ -0,0 +1,343 @@
+
+[RDFLib](https://rdflib.readthedocs.io/en/stable/) is a Python library that provides a simple API for querying RDF data. It is extensible with plugins[^3], allowing it to work with
+different storage backends. In this case, we simply register the Kùzu plugin in RDFLib. For databases like Kùzu that offer on-disk persistence,
+the graph needs to first be opened before we can query it.
+
+```py
+import rdflib
+
+rdflib.plugin.register(
+    "kuzudb",
+    rdflib.store.Store,
+    "graph",
+    "PropertyGraph",
+)
+# Create an RDF graph instance
+graph = rdflib.Graph(store="kuzudb", identifier="kuzudb")
+# Open the RDF graph
+graph.open(configuration="db", create=True)
+```
+
+We can then run a simple SPARQL query via RDFLib to retrieve all triples from the RDF graph.
+
+```py
+import pandas as pd
+
+query = """
+    SELECT ?src ?rel ?dst
+    WHERE {
+        ?src ?rel ?dst .
+    }
+"""
+# Export query results to a Pandas DataFrame
+df = pd.DataFrame([r.asdict() for r in graph.query(query)])
+```
+
+The following result is obtained:
+
+```
+                                src                                              rel                             dst
+0   http://kuzu.io/rdf-ex#Waterloo                       http://kuzu.io/rdf-ex#name                        Waterloo
+1   http://kuzu.io/rdf-ex#Waterloo                 http://kuzu.io/rdf-ex#population                          150000
+2       http://kuzu.io/rdf-ex#Adam                       http://kuzu.io/rdf-ex#name                            Adam
+3       http://kuzu.io/rdf-ex#Adam                        http://kuzu.io/rdf-ex#age                              30
+4    http://kuzu.io/rdf-ex#Karissa                       http://kuzu.io/rdf-ex#name                         Karissa
+5      http://kuzu.io/rdf-ex#Zhang                       http://kuzu.io/rdf-ex#name                           Zhang
+6    http://kuzu.io/rdf-ex#Karissa                     http://kuzu.io/rdf-ex#bornIn  http://kuzu.io/rdf-ex#Waterloo
+7       http://kuzu.io/rdf-ex#Adam                    http://kuzu.io/rdf-ex#livesIn  http://kuzu.io/rdf-ex#Waterloo
+8   http://kuzu.io/rdf-ex#Waterloo  http://www.w3.org/1999/02/22-rdf-syntax-ns#type      http://kuzu.io/rdf-ex#City
+9    http://kuzu.io/rdf-ex#Karissa  http://www.w3.org/1999/02/22-rdf-syntax-ns#type   http://kuzu.io/rdf-ex#student
+10      http://kuzu.io/rdf-ex#Adam  http://www.w3.org/1999/02/22-rdf-syntax-ns#type   http://kuzu.io/rdf-ex#student
+11   http://kuzu.io/rdf-ex#faculty  http://www.w3.org/2000/01/rdf-schema#subClassOf    http://kuzu.io/rdf-ex#person
+12   http://kuzu.io/rdf-ex#student  http://www.w3.org/2000/01/rdf-schema#subClassOf    http://kuzu.io/rdf-ex#person
+13     http://kuzu.io/rdf-ex#Zhang  http://www.w3.org/1999/02/22-rdf-syntax-ns#type   http://kuzu.io/rdf-ex#faculty
+```
+
+## Specify SHACL shape constraints
+
+SHACL allows us to define constraints on the RDF graph. To demonstrate how this works, consider this rather strict
```

Comment:
I think it's OK to use this, but if we will use this, I would be a bit explicit that this is an artificial constraint that's here for the purpose of demonstration: "For purpose of demonstration, consider this simple though somewhat artificial constraint:"